src/osd/scrubber/osd_scrub.cc

```diff
@@ -118,11 +120,68 @@ void OsdScrub::initiate_scrub(bool is_recovery_active)
     }
   }
 
-  auto was_started = select_pg_and_scrub(*env_restrictions);
-  dout(20) << fmt::format(
-		  "scrub scheduling done ({})",
-		  ScrubQueue::attempt_res_text(was_started))
-	   << dendl;
+  // at this phase of the refactoring: minimal changes to the
+  // queue interface used here: we ask for a list of
+  // eligible targets (based on the known restrictions).
+  // We try all elements of this list until a (possibly temporary) success.
+  auto candidates = m_queue.ready_to_scrub(*env_restrictions, scrub_time);
+  auto res{schedule_result_t::none_ready};
```

Comment:
Seems like the entire switch statement below should be replaced by a formatter instance for schedule_result_t and an two methods on schedule_result_t: ```bool schedule_result_t::scheduling_impossible() const``` and ```bool schedule_result_t::scheduling_successful() const```.

Reply:
Not sure I fully understand what you propose above. I will ask for clarifications in our meeting.
But I agree there is no need for that switch-case, and will remove it.
There are 3 possible behaviors acting on the returned result: "initiation went well", "try another target" (later on - even the same PG at the other "deepness level"), or
"stop trying" (if the problem is not specific to the target tried). I'll make a change as a basis for discussion.

Reply:
See the topmost commit for the change